modules/images/image-loading-optimization/optimization.php

```diff
@@ -149,16 +149,15 @@ function ilo_optimize_template_output_buffer( string $buffer ): string {
 	$slug = ilo_get_url_metrics_slug( ilo_get_normalized_query_vars() );
 	$post = ilo_get_url_metrics_post( $slug );
 
-	$url_metrics = $post ? ilo_parse_stored_url_metrics( $post ) : array();
-
-	$needed_minimum_viewport_widths = ilo_get_needed_minimum_viewport_widths(
-		$url_metrics,
-		microtime( true ),
+	$grouped_url_metrics = new ILO_Grouped_URL_Metrics(
+		$post ? ilo_parse_stored_url_metrics( $post ) : array(),
 		ilo_get_breakpoint_max_widths(),
 		ilo_get_url_metrics_breakpoint_sample_size(),
 		ilo_get_url_metric_freshness_ttl()
 	);
 
+	$needed_minimum_viewport_widths = $grouped_url_metrics->get_needed_minimum_viewport_widths( microtime( true ) );
```

Comment:
See my above suggestion of an `is_complete()` method. That could be used here for the `$needs_detection` variable below, replacing the slightly complex logic there and simplifying things. As in:
`$needs_detection = ! $grouped_url_metrics->is_complete()`

Reply:
Nevertheless, the `$needed_minimum_viewport_widths` needs to be passed to below to JavaScript so the client knows whether or not it should try to send a URL metric:

https://github.com/WordPress/performance/blob/4d8baaa822aadc623d0cf222361e2f887e7c3ac3/modules/images/image-loading-optimization/optimization.php#L315-L319

Reply:
Resolving this in favor of the similar conversation as part of https://github.com/WordPress/performance/pull/1010#discussion_r1511735762